packages/hardhat-test-helpers/src/chai/matchers/bigNumber.ts

```diff
@@ -0,0 +1,264 @@
+}
+
+type Methods = "eq" | "gt" | "lt" | "gte" | "lte";
+
+function throwIfUnsafe(i: number) {
+  if (typeof i === "number" && !Number.isSafeInteger(i)) {
+    throw new RangeError(
+      `Cannot compare to unsafe integer ${i}. Consider using BigInt(${i}) instead. For more details, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isSafeInteger`
+    );
+  }
+}
+
+function override(
+  method: Methods,
+  name: string,
+  negativeName: string,
+  utils: Chai.ChaiUtils
+) {
+  return (_super: (...args: any[]) => any) =>
+    overwriteBigNumberFunction(method, name, negativeName, _super, utils);
+}
+
+function normalize(
+  source:
+    | number
+    | bigint
+    | typeof BNType
+    | typeof EthersBigNumberType
+    | typeof BigNumberJsType
+    | string
+): bigint {
+  if (isEthersBigNumber(source)) {
+    return BigInt(source.toString());
+  } else if (isBN(source)) {
+    return BigInt(source.toString());
+  } else if (isBigNumberJsBigNumber(source)) {
+    return BigInt(source.toString(10));
+  } else if (
+    typeof source === "string" ||
+    typeof source === "number" ||
+    typeof source === "bigint"
+  ) {
+    return BigInt(source);
+  } else {
+    throw new Error(`cannot convert ${typeof source} to BigNumber`);
+  }
+}
+
+function isBigNumber(source: any): boolean {
+  return (
+    typeof source === "bigint" ||
+    isEthersBigNumber(source) ||
+    isBN(source) ||
+    isBigNumberJsBigNumber(source)
+  );
+}
+
+function overwriteBigNumberFunction(
+  functionName: Methods,
+  readableName: string,
+  readableNegativeName: string,
+  _super: (...args: any[]) => any,
+  chaiUtils: Chai.ChaiUtils
+) {
+  return function (this: Chai.AssertionStatic, ...args: any[]) {
+    const [actualArg] = args;
+    const expectedFlag = chaiUtils.flag(this, "object");
+    if (chaiUtils.flag(this, "doLength") && isEthersBigNumber(actualArg)) {
+      // TODO: consider whether we really need this case (support for eg
+      // `expect('foo').to.have.length.of.at.least(BigInt(2));`, and if so then
+      // get it tested and generalize it to work with all the different
+      // BigNumbers.
+      _super.apply(this, [actualArg.toNumber()]);
+      return;
+    }
+    function compare(method: Methods, lhs: bigint, rhs: bigint): boolean {
+      if (method === "eq") {
+        return lhs === rhs;
+      } else if (method === "gt") {
+        return lhs > rhs;
+      } else if (method === "lt") {
+        return lhs < rhs;
+      } else if (method === "gte") {
+        return lhs >= rhs;
+      } else if (method === "lte") {
+        return lhs <= rhs;
+      } else {
+        throw new Error(`Unknown comparison operation ${method}`);
+      }
+    }
+    if (isBigNumber(expectedFlag) || isBigNumber(actualArg)) {
+      const expected = normalize(expectedFlag);
+      const actual = normalize(actualArg);
+      throwIfUnsafe(expectedFlag);
+      throwIfUnsafe(actualArg);
```

Comment:
```suggestion
      throwIfUnsafe(expectedFlag);
      throwIfUnsafe(actualArg);
      const expected = normalize(expectedFlag);
      const actual = normalize(actualArg);
```

Reply:
I committed this change, but now I realize I don't like it. :)

The code seems more intuitive with your suggestion, BUT, with the suggestion in place the error message given when trying to use a fractional argument (eg `1.1`) changes from `The number 1.1 cannot be converted to a BigInt because it is not an integer` to `Cannot compare to unsafe integer 1.1. Consider using BigInt(1.1) instead. For more details, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isSafeInteger`.  The message for the "unsafe" case suggests that you wrap the argument in `BigInt(...)`, but that only makes sense if you're already using an integer, and calling `normalize()` first ensures that a non-integer argument raises an error before we attempt to check for integer safety.  Perhaps we could rename the function to `throwIfUnsafeInt()`? Just to be more clear? Just brainstorming...

I'm open to further changes here, but for now I've reverted this suggestion.

Reply:
What if the `normalize` function also checks that the integer is safe? Or is there some scenario where we would call that function with an unsafe integer and that would be ok?

Reply:
I like that, thanks. Addressed in https://github.com/NomicFoundation/hardhat/pull/2428/commits/d875eb1e09dfcf7e66ff3a31a086882e5ae815ad